### ISSUE TITLE
Slow lookups and repeated requests to upstream DNS ("Restarting with new ID.")

### ISSUE THREAD

**Issue by srd424** [systemd/systemd#9785]
**systemd version the issue has been seen with**
> 237 
> (apologies for bending rules, but I suspect Ubuntu won't deal with this in LP)

**Used distribution**
 >Ubuntu 18.04

**Unexpected behaviour you saw**
 > DNS lookups on the LAN are delayed; tcpdumping packets going to the upstream DNS (dnsmasq) shows queries being sent over and over despite replies being received.

**Steps to reproduce the problem**
> Upstream DNS is dnsmasq 2.72.
> systemd-resolve --reset-server-features
> ssh [some-local-host-name]

tcpdump of queries going to to systemd-resolved from local resolver:
```
19:09:59.472197 IP 127.0.0.1.37913 > 127.0.0.53.53: 28655+ A? xu-host.lan. (29)
19:09:59.472272 IP 127.0.0.1.37913 > 127.0.0.53.53: 23544+ AAAA? xu-host.lan. (29)
19:09:59.475742 IP 127.0.0.53.53 > 127.0.0.1.37913: 28655 1/0/0 A 192.168.128.40 (45)
[5 second delay]
19:10:04.477239 IP 127.0.0.1.37913 > 127.0.0.53.53: 28655+ A? xu-host.lan. (29)
19:10:04.480185 IP 127.0.0.53.53 > 127.0.0.1.37913: 28655 1/0/0 A 192.168.128.40 (45)
19:10:04.480377 IP 127.0.0.1.37913 > 127.0.0.53.53: 23544+ AAAA? xu-host.lan. (29)
19:10:04.482609 IP 127.0.0.53.53 > 127.0.0.1.37913: 23544 0/0/0 (29)
```
tcpdump of requests going to upstream DNS:
```
19:13:35.625832 IP 192.168.128.55.46189 > 192.168.128.32.53: 52988+ [1au] A? xu-host.lan. (40)
19:13:35.626112 IP 192.168.128.55.54975 > 192.168.128.32.53: 8958+ [1au] AAAA? xu-host.lan. (40)
19:13:35.628322 IP 192.168.128.32.53 > 192.168.128.55.46189: 52988* 1/0/1 A 192.168.128.40 (56)
19:13:35.628367 IP 192.168.128.32.53 > 192.168.128.55.54975: 8958 0/0/0 (29)
19:13:35.629190 IP 192.168.128.55.54975 > 192.168.128.32.53: 53405+ [1au] AAAA? xu-host.lan. (40)
19:13:35.629858 IP 192.168.128.32.53 > 192.168.128.55.54975: 53405 0/0/0 (29)
19:13:35.630252 IP 192.168.128.55.54975 > 192.168.128.32.53: 1997+ [1au] AAAA? xu-host.lan. (40)
19:13:35.630901 IP 192.168.128.32.53 > 192.168.128.55.54975: 1997 0/0/0 (29)
19:13:35.631223 IP 192.168.128.55.54975 > 192.168.128.32.53: 27245+ [1au] AAAA? xu-host.lan. (40)
19:13:35.631903 IP 192.168.128.32.53 > 192.168.128.55.54975: 27245 0/0/0 (29)
19:13:35.632217 IP 192.168.128.55.54975 > 192.168.128.32.53: 51630+ [1au] AAAA? xu-host.lan. (40)
19:13:35.632920 IP 192.168.128.32.53 > 192.168.128.55.54975: 51630 0/0/0 (29)
19:13:35.633218 IP 192.168.128.55.54975 > 192.168.128.32.53: 16381+ [1au] AAAA? xu-host.lan. (40)
19:13:35.633820 IP 192.168.128.32.53 > 192.168.128.55.54975: 16381 0/0/0 (29)
19:13:35.634135 IP 192.168.128.55.54975 > 192.168.128.32.53: 36615+ [1au] AAAA? xu-host.lan. (40)
19:13:35.634841 IP 192.168.128.32.53 > 192.168.128.55.54975: 36615 0/0/0 (29)
19:13:35.635148 IP 192.168.128.55.54975 > 192.168.128.32.53: 231+ [1au] AAAA? xu-host.lan. (40)
[repeats]
```
Debug logging shows constant restart of transaction due to downgrade of server level:
```
Aug 02 19:13:35 beelink systemd-resolved[6697]: Server doesn't support EDNS(0) properly, downgrading feature level...
Aug 02 19:13:35 beelink systemd-resolved[6697]: Using degraded feature set (UDP) for DNS server 192.168.128.32.
Aug 02 19:13:35 beelink systemd-resolved[6697]: Server feature level is now lower than when we began our transaction. Restarting with new ID.
Aug 02 19:13:35 beelink systemd-resolved[6697]: Transaction 65058 is now 40400.
Aug 02 19:13:35 beelink systemd-resolved[6697]: Cache miss for xu-host.lan IN AAAA
Aug 02 19:13:35 beelink systemd-resolved[6697]: Transaction 40400 for <xu-host.lan IN AAAA> scope dns on lan0/*.
Aug 02 19:13:35 beelink systemd-resolved[6697]: Using feature level UDP+EDNS0 for transaction 40400.
Aug 02 19:13:35 beelink systemd-resolved[6697]: Sending query packet with id 40400.
Aug 02 19:13:35 beelink systemd-resolved[6697]: Processing incoming packet on transaction 40400. (rcode=SUCCESS)
Aug 02 19:13:35 beelink systemd-resolved[6697]: Server feature level is now lower than when we began our transaction. Restarting with new ID.
Aug 02 19:13:35 beelink systemd-resolved[6697]: Transaction 40400 is now 52487.
Aug 02 19:13:35 beelink systemd-resolved[6697]: Cache miss for xu-host.lan IN AAAA
Aug 02 19:13:35 beelink systemd-resolved[6697]: Transaction 52487 for <xu-host.lan IN AAAA> scope dns on lan0/*.
Aug 02 19:13:35 beelink systemd-resolved[6697]: Using feature level UDP+EDNS0 for transaction 52487.
Aug 02 19:13:35 beelink systemd-resolved[6697]: Sending query packet with id 52487.
Aug 02 19:13:35 beelink systemd-resolved[6697]: Processing incoming packet on transaction 52487. (rcode=SUCCESS)
Aug 02 19:13:35 beelink systemd-resolved[6697]: Server feature level is now lower than when we began our transaction. Restarting with new ID.
Aug 02 19:13:35 beelink systemd-resolved[6697]: Transaction 52487 is now 28010.
Aug 02 19:13:35 beelink systemd-resolved[6697]: Cache miss for xu-host.lan IN AAAA
Aug 02 19:13:35 beelink systemd-resolved[6697]: Transaction 28010 for <xu-host.lan IN AAAA> scope dns on lan0/*.
Aug 02 19:13:35 beelink systemd-resolved[6697]: Using feature level UDP+EDNS0 for transaction 28010.
Aug 02 19:13:35 beelink systemd-resolved[6697]: Sending query packet with id 28010.
Aug 02 19:13:35 beelink systemd-resolved[6697]: Processing incoming packet on transaction 28010. (rcode=SUCCESS)
Aug 02 19:13:35 beelink systemd-resolved[6697]: Server feature level is now lower than when we began our transaction. Restarting with new ID.
```

Someone else has found a similar bug and reported it on Launchpad: https://bugs.launchpad.net/ubuntu/+source/systemd/+bug/1771353

Happy to provide pcaps or similar if that would help.

**Comment by srd424** [comment]
Have backported latest dnsmasq (2.79) and am still seeing this issue.

**Comment by srd424** [comment]
systemd-resolve --status: [resolve-status.txt](https://github.com/systemd/systemd/files/2255158/resolve-status.txt)

**Comment by srd424** [comment]
As discussed on IRC, it looks like dnsmasq may not be adding an EDNS0 OPT record to its responses in certain circumstances, which would seem to be a violation of the spec. Still investigating..

**Comment by srd424** [comment]
This dnsmasq commit seems to resolve the issue:

http://thekelleys.org.uk/gitweb/?p=dnsmasq.git;a=commit;h=1682d15a744880b0398af75eadf68fe66128af78

But I think further investigation on the systemd side is probably advisable; the Launchpad bug linked above has a slightly different set of symptoms but seems to be resulting in the same looping behaviour, and older versions of dnsmasq are used widely in embedded devices such as routers.

**Comment by srd424** [comment]
OK, I think I have a simple test case:
```
IFACE=dummy0
SUBNET=10.0.0

ip link add $IFACE type dummy
ifconfig $IFACE ${SUBNET}.1/24
dnsmasq -h -R -d -C /dev/null -2 $IFACE -z -i $IFACE -I lo --host-record=test.test,${SUBNET}.1 &

# now configure systemd-resolved to look at only 10.0.0.1, then

systemd-resolve --reset-server-features
# should exhibit five second delay then connect, assuming sshd is running :)
ssh test.test
```

**Comment by srd424** [comment]
I've finally managed to test this with v239, and I can't reproduce. Would be good to know which commit fixed it though, for cherry picking reasons.
